src/core/Tree/Tree.tsx

```diff
@@ -0,0 +1,265 @@
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+import { CommonProps, useTheme, getFocusableElements } from '../utils';
+import '@itwin/itwinui-css/css/tree.css';
+import cx from 'classnames';
+
+export type TreeContextProps = {
+  nodeDepth: number;
+  subNodeIds?: string[];
+  parentNodeId?: string;
+  groupSize: number;
+  indexInGroup: number;
+};
+
+export const TreeContext = React.createContext<TreeContextProps | undefined>(
+  undefined,
+);
+
+export type NodeData<T> = {
+  /**
+   * Array of the child nodes contained in the node.
+   */
+  subNodes?: Array<T>;
+  /**
+   * Unique id of the node.
+   */
+  nodeId: string;
+  /**
+   * Custom type used to map type `T` to `NodeData`
+   */
+  node: T;
+  /**
+   * Flag whether the node is expanded.
+   */
+  isExpanded?: boolean;
+  /**
+   * Flag whether the node is disabled.
+   */
+  isDisabled?: boolean;
+  /**
+   * Flag whether the node is selected.
+   */
+  isSelected?: boolean;
+  /**
+   * Flag whether the node has sub-nodes.
+   * Used to determine if node should be expandable.
+   */
+  hasSubNodes: boolean;
+};
+
+type FlatNode<T> = {
+  nodeProps: NodeRenderProps<T>;
+  depth: number;
+  subNodeIds?: Array<string>;
+  parentNode?: FlatNode<T>;
+};
+
+export type NodeRenderProps<T> = Omit<NodeData<T>, 'subNodes'>;
+
+export type TreeProps<T> = {
+  /**
+   * Custom renderer for items inside tree, recommended to use `TreeNode` component.
+   */
+  nodeRenderer: (props: NodeRenderProps<T>) => JSX.Element;
+  /**
+   * Array of custom data used for `TreeNodes` inside `Tree`.
+   */
+  data: T[];
+  /**
+   * Function to map custom type `T` to `NodeData`, which will be used to render a tree node in `nodeRenderer`.
+   */
+  getNode: (node: T) => NodeData<T>;
+} & Omit<CommonProps, 'title'>;
+
+/**
+ * Tree component used to display a hierarchical structure of `TreeNodes`. 
+ * User should control state of expanded, selected and disabled nodes using `getNode` prop.
+ * @example
+  type DemoData = {
+    id: string;
+    label: string;
+    subItems: DemoData[];
+  };
+
+  const data: Array<DemoData> = [
+    {
+      id: 'Node-1',
+      label: 'Facility 1',
+      subItems: [{ id: 'Node-1-1', label: 'Unit 1', subItems: [] }],
+    },
+    {
+      id: 'Node-2',
+      label: 'Facility 2',
+      subItems: [{ id: 'Node-2-1', label: 'Unit 2', subItems: [] }],
+    },
+  ];
+
+  const [expandedNodes, setExpandedNodes] = React.useState<Array<string>>([]);
+  const onNodeExpanded = (nodeId: string, isExpanded: boolean) => {
+    if (isExpanded) {
+      setExpandedNodes((oldExpanded) => [...oldExpanded, nodeId]);
+    } else {
+      setExpandedNodes((oldExpanded) =>
+        oldExpanded.filter((item) => item != nodeId),
+      );
+    }
+  };
+
+  const getNode = (node: DemoData): NodeData<DemoData> => {
```

Comment:
need to review again this code sample.. I think here node type is wrong

Reply:
Changed a bit

Reply:
nodeRenderer I guess also can wrap in useCallback as in story.. and should be ok